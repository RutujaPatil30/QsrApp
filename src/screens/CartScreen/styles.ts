import { StyleSheet } from "react-native";
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';
import theme from "../../utils/themes";


const styles = StyleSheet.create({
    container:{
        flex:1
    },
    topView: {
        height: hp('18%'),
        backgroundColor: theme.colors.secondary,
        padding: hp('2%'),
        borderBottomRightRadius: wp('10%'),
        borderBottomLeftRadius: wp('10%'),
    },
    pageTitle: {
        fontSize: theme.fonts.detailsLabel,
        color: theme.colors.primary,
        fontWeight: 'bold',
    },
    scrollViewStyle: {
        flexGrow:1,
        position: 'absolute',
        top: hp('4%'),
        height: hp('100%'),
        padding: hp('2%'),
        borderRadius: wp('5%'),
    },
    totalItemViewStyle: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginHorizontal: hp('2%'),
    },
    totalItemTextStyle: {
        color: theme.colors.black,
        fontSize: theme.fonts.subtitle
    },
    totalPriceText: {
        color: theme.colors.black,
        fontWeight: 'bold',
        fontSize: theme.fonts.labels
    },
    gratisStyle: {
        marginBottom:wp('3%') ,
        fontSize: theme.fonts.subtitle,
        color: theme.colors.black
    },
    buttonContainer :{
        alignItems: "center",
        justifyContent:"flex-end", 
        top:hp('70%')
    },
    bottomContainer1: {
        color: theme.colors.secondary,
        fontSize: 26,
        fontWeight: 'bold',
        marginLeft: 20,
        marginTop: 15,
    },
    bottomContainer2: {
        color: theme.colors.primary,
        fontSize: 18,
        fontWeight: '500',
    },
    noRideContainer: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    noRideImage: {
        width:wp('100%'),
        height: hp('20%'),
        resizeMode: 'contain',
      },
})

export default styles;