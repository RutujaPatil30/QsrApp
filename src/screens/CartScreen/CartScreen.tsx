import React, { useEffect, useState } from "react";
import { View, Text, ScrollView, Image, TouchableOpacity } from "react-native";
import CustomButton from "../../components/CustomButton/CustomButton";
import CustomCartElement from "../../components/CustomCartElement/CustomCartElement";
import baselocalization from "../../utils/baselocalization";
import theme from "../../utils/themes";
import styles from "./styles";
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';
import { useDispatch, useSelector } from "react-redux";
import { images } from "../../utils/images";
import RazorpayCheckout from 'react-native-razorpay';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { addRecentOrder, clearCart } from "../../redux/action";

const CartScreen = (props: any) => {
    const { navigation } = props;
    const dispatch = useDispatch();
    const cartItems = useSelector((state: any) => state.reducer.cartItems);
    const userDetails = useSelector((state: any) => state.reducer.currentUser);
    const userId = useSelector((state: any) => state.reducer.currentUser.id);
    const [totalItems, setTotalItems] = useState(0);
    const [totalPrice, setTotalPrice] = useState(0);
    const [isEmpty, setIsEmpty] = useState(true);

    useEffect(() => {
        let count = 0;
        cartItems.forEach(item => {
            count = 1;
            setIsEmpty(false);
        });
        if (count == 0) {
            setIsEmpty(true);
        }
    }, [cartItems]);

    useEffect(() => {
        let newTotalItems = 0;
        let newTotalPrice = 0;

        cartItems.forEach(item => {
            newTotalItems += item.qty;
            newTotalPrice += item.qty * item.price;
        });
        setTotalItems(newTotalItems);
        setTotalPrice(newTotalPrice);
    }, [cartItems]);

    const PayNow = () => {
        var options = {
            description: "Cart Items",
            image: 'https://andhra.mallsmarket.com/sites/default/files/styles/large/public/images/brands/McDonalds-Logo.jpg',
            currency: 'INR',
            key: 'rzp_test_6BL5V2w5EMP1xC',
            amount: totalPrice * 100,
            name: `McDonald's`,
            order_id: '',
            prefill: {
                email: userDetails.email,
                contact: userDetails.phoneNumber,
                name: userDetails.name,
            },
            theme: { color: theme.colors.primary },
        };
        RazorpayCheckout.open(options)
            .then(data => {
                alert(`Success: Payment Successful`);
                const orderDetails = {
                    items: cartItems.map(item => ({
                        name: item.title, // Ensure 'title' is the correct field for the product name
                        price: item.price,
                        qty: item.qty
                    })),
                    totalPrice: totalPrice
                };
                dispatch(addRecentOrder(orderDetails));
                
                dispatch(clearCart()); 
            })
            .catch(error => {
                alert(`Error: ${error.code} | ${error.description}`);
            });
    }

    return !isEmpty ? (
        <View>
            <View style={styles.topView}>
                <Text style={styles.pageTitle}>{baselocalization.cartScreen.cartScreenTitle}</Text>
                <TouchableOpacity onPress={()=>{navigation.navigate('ProfileScreen')}}>
                    <Icon name="account-circle" size={35} color={theme.colors.primary} style={{ marginLeft: wp('55%') }} />
                </TouchableOpacity>
            </View>
            <ScrollView style={styles.scrollViewStyle} >
                {cartItems.map(item =>
                    <View style={{ margin: 20 }}>
                        <CustomCartElement
                            testID={"CartScreen_CartProduct"}
                            itemImage={item.image}
                            itemName={item.title}
                            itemPrice={item.price * item.qty}
                            itemQty={item.qty}
                            itemId={item.id} />
                    </View>
                )}

                <View style={styles.totalItemViewStyle}>
                    <View>
                        <Text style={styles.totalItemTextStyle}>
                            {baselocalization.cartScreen.cartScreenTotalItem}
                        </Text>
                        <Text style={styles.gratisStyle}>
                            {baselocalization.cartScreen.cartScreenBiayaPengiriman}
                        </Text>
                        <Text style={styles.totalPriceText}>
                            {baselocalization.cartScreen.cartScreenTotalBiaya}
                        </Text>
                    </View>
                    <View style={{ alignItems: 'flex-end' }}>
                        <Text style={styles.totalItemTextStyle}>{totalItems}</Text>
                        <Text style={styles.gratisStyle}>
                            {baselocalization.cartScreen.cartScreenGratis}
                        </Text>
                        <Text testID='CartScreen_TotalPrice' style={styles.totalPriceText}>Rs {totalPrice}</Text>
                    </View>
                </View>
            </ScrollView>
            <View style={styles.buttonContainer}>
                <CustomButton testID='CartScreen_NextButton' text={baselocalization.cartScreen.cartScreenPayButton}
                    disabled={false}
                    buttonStylesProps={{
                        backgroundColor: theme.colors.primary,
                        borderColor: theme.colors.primary,
                        textColor: theme.colors.white,
                        width: wp('45%'),
                        height: hp('6%'),
                        fontSize: theme.fonts.labels
                    }}
                    onPress={() => { PayNow() }}
                />
                <View style={{ marginBottom: 10 }}></View>
                <CustomButton testID='CartScreen_PayButton' text={baselocalization.cartScreen.cartScreenButtonText}
                    disabled={false}
                    buttonStylesProps={{
                        backgroundColor: theme.colors.primary,
                        borderColor: theme.colors.primary,
                        textColor: theme.colors.white,
                        width: wp('45%'),
                        height: hp('6%'),
                        fontSize: theme.fonts.labels
                    }}
                    onPress={() => { navigation.navigate("CardScreen") }}
                />
            </View>
        </View>
    ) : (
        <View style={styles.container}>
            <View style={styles.topView}>
                <Text style={styles.TextContainer}>
                    {baselocalization.cartScreen.cartScreennoItem}
                </Text>
            </View>
            <View style={styles.noRideContainer}>
                <>
                    <Image source={images.cartScreen_emptyCart} style={styles.noRideImage} />
                </>
            </View>
        </View>
    )
}

export default CartScreen;