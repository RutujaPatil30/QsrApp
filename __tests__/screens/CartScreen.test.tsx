import React from 'react';
import { render, screen } from '@testing-library/react-native';
import CartScreen from '../../src/Screens/CartScreen/CartScreen';
import '@testing-library/jest-native/extend-expect';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

const mockStore = configureStore([]);

const initialState = {
    reducer: {
        cartItems: [
            {
                id: 9,
                title: 'McFlurry Oreo',
                description:
                    'Delicious soft serve meets crumbled oreo cookies, a match made in dessert heaven.',
                image: require('../../src/assets/breakfast.png'),
                price: 10.0,
                qty: 1,
            },
            {
                id: 10,
                title: 'Hot Fudge',
                description:
                    'A sinful delight, soft serve topped with delicious, gooey hot chocolate fudge.',
                image: require('../../src/assets/breakfast.png'),
                price: 15.0,
                qty: 1,
            },
        ],
        totalItems: 2,
        totalPrice: 25.0,
    },
};

const store = mockStore(initialState);

describe('Checking Cart Screen', () => {
    it('should render the expected UI elements', () => {
        render(
            <Provider store={store}>
                <CartScreen />
            </Provider>
        );

        const cartProduct = screen.queryAllByTestId("CartScreen_CartProduct");
        const cartButton = screen.getByTestId('CartScreen_NextButton');

        expect(cartProduct.length).toBeGreaterThan(0);
        expect(cartButton).toBeTruthy();
    });

    it('should display the correct total price', () => {
        render(
            <Provider store={store}>
                <CartScreen />
            </Provider>
        );

        const totalPriceText = screen.getByTestId('CartScreen_TotalPrice');
        expect(totalPriceText.props.children.join('')).toBe('Rs 25');
    });

});

describe('CartScreen Snapshot Test', () => {
    it('renders correctly', () => {
        const tree = renderer.create(
            <Provider store={store}>
                <CartScreen />
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});