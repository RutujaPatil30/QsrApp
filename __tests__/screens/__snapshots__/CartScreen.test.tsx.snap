// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartScreen Snapshot Test renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffc72c",
        "borderBottomLeftRadius": 75,
        "borderBottomRightRadius": 75,
        "height": 240,
        "padding": 26.5,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffc72c",
          "fontSize": 26,
          "fontWeight": "bold",
          "marginLeft": 20,
          "marginTop": 15,
        }
      }
    >
      Pesananmu
    </Text>
    <Text
      style={
        {
          "color": "#da291c",
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      No Item in the cart !!!
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <RCTImageView
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      internal_analyticTag={null}
      resizeMode="contain"
      source={
        [
          {
            "height": undefined,
            "uri": undefined,
            "width": undefined,
          },
        ]
      }
      style={
        {
          "height": 267,
          "overflow": "hidden",
          "resizeMode": "contain",
          "width": 750,
        }
      }
    />
  </View>
</View>
`;
